### DISCUSSION TITLE
Websocket

### DISCUSSION THREAD

**Discussion by massimoliani** [directus/directus#9129] [Archive]
I'm developing a project that automatically publish to a WordPress web site.
I implement a plugin that wait for a message on a websocket from the direct us server to update the blog page in case of new post.
I implement the websocket on a different port than the server port as i can't find a way to access express app to register the websocket server on the same port has the regular direct us server.

Is there any way to access the express app and register the websocket server using the npm module for express?

Thank you

**Comment by rijkvanzanten** [comment]
You should be able to use the init hooks to tap into the express server instance itself 👍🏻

**Reply by massimoliani** [reply]
Thanks I'll check the direct us sources to find the parameters passed to the hook.

**Reply by massimoliani** [reply]
Hi @rijkvanzanten I post my peronal implementation of websocket to share the solution to anyone would like to implement it.

I test npm express-ws but can't work well, npm ws work fine.

Here the cose inside server.start.before hook:
```js
    "server.start.before": function ({ server }) {
      const ws = require("ws");
      const wssWP = new ws.WebSocket.Server({
        server: server,
        path: "/path_to_ws",
      });

      const file = require("somewhore/somefile.js");
      file.initWS(wssWP);

      console.log("Websocket created");
    },
```

Inside initWSS:
```js
var initWS = function (wss) {
  wss.on("connection", function (ws) {
    ws.on("message", function incoming(message) {
      var data = JSON.parse(message);
      switch (data.type) {
        case "open":
          console.log("Client socket open message:" + data.status);
          break;
        case "ping":
          console.log("Client socket ping message:" + data.status);
          ws.send(JSON.stringify({ status: data.status, type: "pong" }));
          break;
        default:
          console.log(
            "Client message type not handled payload:" + JSON.stringify(data)
          );
          break;
      }
    });

    console.log("New socket connection request");
    ws.send(JSON.stringify({ status: "ok", type: "connection" }));
    if (global.wp.ws == null) {
      global.wp.ws = [];
    }
    var idx = global.wp.ws.findIndex((e) => e == ws);
    if (idx == -1) {
      global.wp.ws.push(ws);
    }
  });
};
```
the socket send and receive a JSON object type/status member.

Is possible to place the websocket in a path different from the root and link to the same port of main server.

**Reply by rijkvanzanten** [reply]
That's awesome!! @Nitwel must be drooling right now 🤤

**Reply by benhaynes** [reply]
Any talk of `ws` certainly has _me_ drooling!! I'd love to get that support in Directus. ❤️
